File: frontend/src/lib/api/stock.ts
// 株価 OHLC API（`backend/routers/stock.py`）の薄い型付きラッパ。

import { api } from '@/lib/api/client';

export type OhlcPeriod = '1mo' | '3mo' | '6mo' | '1y' | '2y';

export interface OhlcBar {
  time: string; // YYYY-MM-DD
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

// テクニカル指標のクロスイベント（🆕、ゴールデンクロス/デッドクロス・MACDクロス）。
// チャート上のマーカー描画・hoverツールチップに使う。
export type ChartEventKind = 'golden_cross' | 'dead_cross' | 'macd_bullish_cross' | 'macd_bearish_cross';

export interface ChartEvent {
  date: string; // YYYY-MM-DD
  kind: ChartEventKind;
  label: string;
}

export interface OhlcResponse {
  bars: OhlcBar[];
  events: ChartEvent[];
}

export function fetchOhlc(symbol: string, period: OhlcPeriod = '6mo'): Promise<OhlcResponse> {
  return api.get<OhlcResponse>(`/stock/${encodeURIComponent(symbol)}/ohlc?period=${period}`);
}

// 🆕 P26: ポートフォリオの買い/売りフォームが開いた時点の最新値を初期値にするための
// 軽量エンドポイント。取得失敗時は price 等が null で返る（フォームは手入力にフォールバック）。
export interface Quote {
  symbol: string;
  price: number | null;
  prev_close: number | null;
  change_pct: number | null;
}

export function fetchQuote(symbol: string): Promise<Quote> {
  return api.get<Quote>(`/stock/${encodeURIComponent(symbol)}/quote`);
}

export interface StockNote {
  code: string;
  note_title: string;
  content: string;
}

// ナレッジベース（Vault）ノートを本文込みで取得する。UI 表示専用（ポップアップ）で
// LLM プロンプトへは使わない（`backend/services/vault/brand_notes_service.get_raw_note_content`
// docstring 参照）。ノート未整備の銘柄では `null` を返す（エラーではない）。
export function fetchStockNote(symbol: string): Promise<StockNote | null> {
  return api.get<StockNote | null>(`/stock/${encodeURIComponent(symbol)}/note`);
}

// 🆕 P26: 証券コード/銘柄名の部分一致検索（ポートフォリオへの銘柄追加用）。
export interface TickerInfo {
  code: string;
  name: string;
  sector: string | null;
}

export function searchStocks(query: string): Promise<TickerInfo[]> {
  return api.get<TickerInfo[]>(`/stock/search?q=${encodeURIComponent(query)}`);
}

File: backend/models/stock.py
"""株価 OHLC のスキーマ定義（銘柄詳細画面のチャート用、🆕 P8c）."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict


class OhlcBar(BaseModel):
    """ローソク足 1 本分（`lightweight-charts` のフォーマットに合わせる）.

    `time` は日足（interval="1d"）では `YYYY-MM-DD` 文字列、分足（🆕、interval="60m"/"15m"）
    では UTC 起点の Unix 秒（`int`）。`lightweight-charts` の `Time` 型が「日付文字列は
    1日単位の解像度までしか表現できない」制約を持つため、同日内に複数本並ぶ分足はタイム
    スタンプで表現する必要がある（`routers/stock.py` 参照）。
    """

    model_config = ConfigDict(frozen=True)

    time: str | int
    open: float
    high: float
    low: float
    close: float
    volume: float


class ChartEvent(BaseModel):
    """テクニカル指標のクロスイベント 1 件（🆕、ローソク足チャートのマーカー表示用）.

    `services/scoring/technical_analysis.detect_cross_events` の結果をそのまま公開する。
    """

    model_config = ConfigDict(frozen=True)

    date: str  # YYYY-MM-DD
    kind: Literal["golden_cross", "dead_cross", "macd_bullish_cross", "macd_bearish_cross"]
    label: str


class OhlcResponse(BaseModel):
    """`GET /api/stock/{symbol}/ohlc` のレスポンス（🆕、四本値＋兆候イベント）."""

    model_config = ConfigDict(frozen=True)

    bars: list[OhlcBar]
    events: list[ChartEvent]


class Quote(BaseModel):
    """単一銘柄の直近値（🆕 P26、ポートフォリオの買い/売りフォームで取引時点の最新値を
    初期値に使うための軽量エンドポイント用）."""

    model_config = ConfigDict(frozen=True)

    symbol: str
    price: float | None
    prev_close: float | None
    change_pct: float | None
